Chore: Advanced Bloom Filter Deprecations

bbazukun123 opened the pull request:
- Add deprecations for the Advanced Bloom filter.
- Use `GpuProgram.from` and `GlProgram.from` for caching.

bigtimebuddy commented on src/advanced-bloom/AdvancedBloomFilter.ts:
```diff
@@ -73,11 +99,35 @@ export class AdvancedBloomFilter extends Filter
+    constructor(options?: Partial<DeprecatedAdvancedBloomFilterOptions>);
+    constructor(options?: AdvancedBloomFilterOptions | Partial<DeprecatedAdvancedBloomFilterOptions>)
     {
+        if (typeof options?.pixelSize === 'number' || Array.isArray(options?.pixelSize))
+        {
+            deprecation(v8_0_0, 'AdvancedBloomFilterOptions.pixelSize now only accepts {x, y} PointData type.');
+            options.pixelSize = convertDeprecatedPixelSize(options.pixelSize);
+        }
+
```
Because we are doing `Object.assign(this, options)`, you'll get two deprecation messages here. This is redundant if handled already in pixelSize setter.

bigtimebuddy commented on src/advanced-bloom/AdvancedBloomFilter.ts:
```diff
@@ -1,10 +1,36 @@
 
+type DeprecatedPixelSizeValue = PointData | number[] | number;
+
+interface DeprecatedAdvancedBloomFilterOptions
+{
+    threshold: number,
+    bloomScale: number,
+    brightness: number,
+    kernels: number[] | null,
+    blur: number,
+    quality: number,
+    pixelSize: DeprecatedPixelSizeValue,
+    resolution: number,
+}
+
```
This is a pretty verbose way to deprecate `pixelSize` type. Another approach is to change the name. For instance:

```ts
export interface AdvancedBloomFilterOptions {
    size?: PointData,
    /** @deprecated use size instead */
    pixelSize?: PointData | number[] | number,
}
```
This way you don't duplicate options. With renaming you can do clean-breaks on the `pixelSize`, `pixelSizeX`, `pixelSizeY` setters/getters. We want to make sure that `@deprecated` doc tag is added so that it shows up in user's IDE. E.g., **filter.~~pixelSizeX~~ = 10**

If you don't want to rename, you can do something like this:

```ts
interface DeprecatedAdvancedBloomFilterOptions extends Omit<AdvancedBloomFilterOptions, 'pixelSize'> {
    /** @deprecated use PointData only */
    pixelSize?: PointData | number[] | number;
}
```
At least this way you don't duplicate the interface.

bigtimebuddy commented on src/advanced-bloom/AdvancedBloomFilter.ts:
```diff
@@ -158,42 +208,68 @@ export class AdvancedBloomFilter extends Filter
     set pixelSizeY(value: number) { this._blurFilter.pixelSizeY = value; }
 }
+
+function convertDeprecatedPixelSize(value: DeprecatedPixelSizeValue): PointData
```
I would inline this function in the pixelSize setter since you don't need to use it in the constructor.

bigtimebuddy commented on src/advanced-bloom/AdvancedBloomFilter.ts:
```diff
@@ -73,11 +99,35 @@ export class AdvancedBloomFilter extends Filter
+     * @param {number|number[]|PIXI.PointData} [options.pixelSize=1] - the pixelSize of the Blur filter.
+     * @param {number} [options.resolution=PIXI.settings.FILTER_RESOLUTION] - The resolution of the Blur filter.
+     */
+    constructor(options?: Partial<DeprecatedAdvancedBloomFilterOptions>);
+    constructor(options?: AdvancedBloomFilterOptions | Partial<DeprecatedAdvancedBloomFilterOptions>)
```
Why is there a mix of using `Partial<>` here and not? These interfaces are more useful if we keep the properties optional instead of using Partial so that users can cast:

```ts
const options = { pixelSize: 10 } as DeprecatedAdvancedBloomFilterOptions;
```

bigtimebuddy reviewed (changes requested):
I left some feedback but overall, I don't think we need this change. What is the harm in keeping `number` and `number[]` as a type for pixelSize and just convert it to a PointData in the pixelSize setter. Then you don't need all this extra work.

bigtimebuddy commented on src/advanced-bloom/AdvancedBloomFilter.ts:
```diff
@@ -73,11 +99,35 @@ export class AdvancedBloomFilter extends Filter
-    constructor(options?: AdvancedBloomFilterOptions)
+    constructor(options?: AdvancedBloomFilterOptions);
+    /**
+     * @deprecated since 8.0.0
```
This should be 6.0.0, not 8.0.0.

bbazukun123 commented:
> I left some feedback but overall, I don't think we need this change. What is the harm in keeping `number` and `number[]` as a type for pixelSize and just convert it to a PointData in the pixelSize setter. Then you don't need all this extra work.

Thanks Matt, and apologies for the mess. I have addressed the feedback - essentially reverting the deprecation and allowing the number[]/[] as possible types for pixelSize param. Hope this looks neater now 🤞

bigtimebuddy reviewed (approved):
Much better!
